### DISCUSSION TITLE
Importing a Linked Node in Another Map / Syncing Nodes Across Maps

### DISCUSSION THREAD

**Discussion by CD80-86** [freeplane/freeplane#1706] [Help] [answered]
Hello! I am long-time user of freeplane for study in medical school, with almost no coding background. I've managed to write a few very basic scripts myself but need some help on this particular issue.

**Basic Goal**
In Map A, I have a node N1.
In Map B, I have a node N2.

N1 has a global hyperlink to N2.

From N1, I would like to be able to, in one click / with a script, copy N2 and paste it (and any children) as a child of N1 - let's call this new copy N2x.

_What I have tried:_
The "Import Linked Branch" built into Freeplane does not work on linked nodes that are located in another map different to the one you are trying to import to (the freeplane API makes reference to this as well).

I tried to write a script myself using _appendBranch_ but I really don't have anywhere near enough experience to do something this complex - I know almost nothing about java or groovy!

I would really appreciate if someone could help me write a script to accomplish this goal - effectively a "paste linked node from another map" feature. I would execute the script on N1 in Map A, it would follow the link in N1 to N2 in Map B, and then copy and paste [nB without deleting the original N2] it as a child of N1 in Map A.

**Bonus Goal**
Ideally, I would like the script to act as a "syncing" function: 
say I've created N2x as a child to N1 by running the script once before, and at a later date I have edited the original N2 to "N2.1" in Map B. I would like for the script, when I run it a second time on N1, to replace/overwrite the copy (N2x) so that it reflects the changes present in N2.1. 
This way I can run the script on Map A to update any changes made in Map B to be reflected in Map A, without having to delete the old/obsolete version of N2, which will be very time consuming with the scale I'm planning to use this script on.

**Bonus Bonus Goal**
The cherry on top would be a separate script that executes on all nodes in Map A so I can choose to sync the whole map immediately, but I think I can figure this one out on my own. 

My apologies if this kind of thing has been addressed in the past, but I searched this discussion thread and the old one at sourceforge, as well as browsing the API to see if I could figure out a simple way to do it, but I couldn't find an answer. 

**Thank you** in advance to anyone who reads this and can give any advice or help!

**Comment by euu2021** [comment]
I will try to create the script, tomorrow, but what immediately comes to my mind is: have you considered using one big single map? I have difficulty in visualizing any advantages of using multiple maps. This setup that you want is much easier when done inside the same map.

**Reply by CD80-86** [reply]
Amazing - thank you very, very much! 

**TL;DR** is one big map would be too slow due to a large number of pasted images in the map.

The reason I can't use a single map is because "Map B" as I described it above contains medical school notes with lots and lots of pasted images - I store them all in one folder on my desktop and so far they represent ~2GB. 
As maps got large, I started to experience slowdowns when moving nodes or resizing images, so I actually have many "Map B"s, roughly corresponding to different medical specialties, each containing very many nodes corresponding to specific conditions (e.g. the map "Respiratory Medicine" contains the condition/node "Pneumonia").

"Map A" in this case is going to be a map where I list different medical "problems/symptoms" (e.g. "patient has a cough"), and I will among other things be listing conditions that can cause each problem. The nodes in this list will each contain a link to the corresponding node in Map B that actually contains all the content of the condition listed. In reality, each condition in Map B has a child node called "Summary", which is less detailed and contains no images, and this is the only node I want to copy across and sync. 

Because this is all quite long-winded and complicated to explain, I simplified the case down for my request. Having made some attempts so far through trial and error, I think I will be able to modify any final script to point to and copy the Summary node instead of the whole parent condition node. It's quite rewarding figuring out how to do things, but the task of handling links/URIs, map access operations, and clipboard operations from scratch was just beyond my capabilities.

Can I offer you a 'thank you' gift if you manage to write the script? I want to reimburse you for the time you spend on it.

**Reply by euu2021** [reply]
I found that it was discussed at #1024.

For now, what I have:
- this script creates links between maps: https://docs.freeplane.org/scripting/Scripts_collection.html#insert-link-to-a-node-in-another-map
- in that link, there is the node ID of the target node

So, the next step is to find a practical way to open the new map and navigate to the node. I will continue, later.

> TL;DR is one big map would be too slow due to a large number of pasted images in the map.

Good to know that images hurt performance. My map doesn't have this amount of images, so I never noticed this problem.

> Can I offer you a 'thank you' gift if you manage to write the script? I want to reimburse you for the time you spend on it.

It's not necessary, thanks. I think this feature would be of general interest, so the time spent on it is time well spent.

**Reply by euu2021** [reply]
I managed to do Read operations, in the background (headless load), with nodes in another map. For example, this retrieves the text of the node:

```
targetMap = c.mapLoader('C:/Users/kauen/Downloads/Map 2.mm').load()
targetNode = targetMap.root.find{it.id == 'ID_1216987901'}
targetNode.text
```

But, I couldn't make write operations. For example, trying to set the text of the node gives me this exception:
`java.lang.IllegalArgumentException: argument type mismatch`

🤔🤔🤔

**Reply by euu2021** [reply]
Oh, the problem was that `targetNode` was returning an Array. Now it works:

```
targetMap = c.mapLoader('C:/Users/kauen/Downloads/Map 2.mm').load()
targetNode = targetMap.root.find{it.id == 'ID_1216987901'}[0]
targetNode.text = "aaa"
```

I think this script is gonna fly 🚀😮. The pieces are there. Now, it's a matter of assembling.

**Reply by CD80-86** [reply]
Thank you, this is so helpful. I've pieced together a script that does the job as far as getting to the stage of _attempting to_ copy the source node:

```
        if (node.Link.getUri() == null){
        return null
        }
    
        else { //ALL of the above code is to avoid any exceptions if the selected node has no link
        
        def SourceURI = node.Link.getUri() //returns node's Link's URI
        def SourceStr = SourceURI.toString() //converts URI to string
        def SourceStrNoFile = SourceStr.getAt(6..(SourceStr.length()-1))
        
        def SourceIDStartPosition = SourceStrNoFile.indexOf('ID', 0)
        
        //defines SourceID as the ID of the source node
        def SourceID =  SourceStrNoFile.getAt(SourceIDStartPosition..(SourceStrNoFile.length()-1)) 
        //defines SourcePath as the path of the source map 
        def SourcePath = SourceStrNoFile.getAt(0..(SourceIDStartPosition-2)) //
       
       //Defines and Loads SourceMap
       def SourceMap = c.mapLoader(SourcePath).load()
       //Defines SourceNode
       def SourceNode = SourceMap.root.find{it.id == (SourceID)}[0]
       
       SourceNode.clipboardController.copy()
    }
```

I'm assuming there's a much more elegant way to separate the URI into map path and node ID, but this works!

However, the last line of code (which I found by browsing the source code - _org.freeplane.features.map.clipboard.clipboardController_) fails to act on "SourceNode" in the other map, it just executes on the currently selected node in the current map. In fact, I can't even point it to other nodes in the current map - it always goes to the currently selected node.

Therefore, I suppose I need a way to tell freeplane to perform operations on the external map as if it was the currently open map (like telling freeplane to open the external map and select the relevant node 'in the background', then switch back), but I don't know how to do this as I can't find any documentation on this aspect of freeplane's code. I'm assuming it's quite complicated.

I am guessing an alternative is to get all the properties of the SourceNode, without going via the clipboard, but I want to copy across not just the SourceNode's text, but also its formatting, its children, their formatting, and any links or pasted images etc. I have no idea how I'd do this with code - I'm assuming it's also relatively complicated unless there's a simple way to say "get everything about this node and its children" and then paste that.

Any further help would be appreciated!

**Reply by abc16361** [reply]
You can try this to see whether it works for you,


```
// @ExecutionModes({ON_SINGLE_NODE})


def nodeLink = node.link.text
def nLnk = nodeLink.split('#')
def mapPath = nLnk[0].replace('freeplane:/%20/', '').replace('file:///','').replace('file:/','')
def nid = nLnk[1]

def sourceMap = c.mapLoader(mapPath).withView().load()
def sourceNode = sourceMap.node(nid)
c.select(sourceNode)
menuUtils.executeMenuItems(['CopyAction'])
menuUtils.executeMenuItems(['CloseAction'])
menuUtils.executeMenuItems(['PasteAction'])
```

**Reply by abc16361** [reply]
I add some lines and this new code can update the node copy to the source node. It is not a sync but a replacement update since it delete the node copy first before doing copy and paste.

```
// @ExecutionModes({ON_SINGLE_NODE})


def nodeLink = node.link.text
def nLnk = nodeLink.split('#')
def mapPath = nLnk[0].replace('freeplane:/%20/', '').replace('file:///','').replace('file:/','')
def nid = nLnk[1]
def childnode = node.find{ it.attributes.getNames().contains(nid) }
if(childnode){
	c.select(childnode)
	c.selected.delete()
}

def sourceMap = c.mapLoader(mapPath).withView().load()
def sourceNode = sourceMap.node(nid)
c.select(sourceNode)
c.selected[nid] = ''
menuUtils.executeMenuItems(['CopyAction'])
c.selected.attributes.removeAll(nid)
menuUtils.executeMenuItems(['SaveAction'])
menuUtils.executeMenuItems(['CloseAction'])

menuUtils.executeMenuItems(['PasteAction'])
```

**Reply by euu2021** [reply]
It's possible to use appendBranch across maps. For example, using the terminology from the first post, if I have n2 selected and use this script, the branch is copied to n1 (ID_1216987901), located at map A (C:/Users/kauen/Downloads/Map 2.mm):

```
n2 = node //define n2
mapA = c.mapLoader('C:/Users/kauen/Downloads/Map 2.mm').load() //load map A in the background
n1 = mapA.root.find{it.id == 'ID_1216987901'}[0] //define n1
n1.appendBranch(n2)
```

Using appendBranch is different of using copy and paste, because it preserves the node ID, which may be useful when creating the sync feature.

Now, the script by abc16361 does the job of treating the link on node 2, but I still need to understand how to use the path of the maps (relative and absolute). For example, in the example above, you can see that my map is called "Map 2.mm", but in the link shows up as "Map%202.mm". Probably we need to adapt the "Insert Link to a node in another map" script, in order to make it more convenient for that situation.

**Reply by CD80-86** [reply]
I've managed to combine all of the information you've each provided and written script that works, based on the method of calling copy and paste actions. 

In addition, I've added some code to deal with situations where SourceMap may already be open in the current instance of freeplane, such as if you're editing it on the fly:

- If SourceMap is not already open, then the script will close its background instance on completion
- If SourceMap is already open, then the script will keep it open and just switch back to TargetMap

I did this by collecting the names of open maps and comparing it to the name of SourceMap in its file a

@EdoFro 
After some experimenting, it seemed that copying and pasting _does_ preserve node IDs across maps, though of course that will stop being the case as soon as you paste the node into a map that already contains a node with the same ID, in which case the pasted node will be assigned an entirely new ID. Nevertheless, maybe there are further benefits in using appendBranch instead.

Your point about how spaces are handled in the link is excellent - I hadn't thought of that and managed to add a line to handle it. It doesn't seem like any other characters will cause issues, because freeplane doesn't let you save maps with special characters in their file name.

Here is the working script:

     // @ExecutionModes({ON_SINGLE_NODE})

    //Will return null (instead of an exception) if selected node has no link
    if (node.Link.getUri() == null){
        return null
    }
    else { 

        //Below is Set-Up for returning to TargetMap near the end, before pasting
	//Below Finds the Currently Selected Node's ID# as a string
	def rawTargetID = node.getId()
	    rawTargetID.toString()
	def TargetNodeID = rawTargetID.getAt(3..(rawTargetID.length()-1))

	//Below Finds the Currently Selected Node's (Target's) map path as a string
	def rawTargetURI = node.mindMap.file.absoluteFile.toURI()	
	def rawTargetPath = rawTargetURI.toString()

	def TargetMapPath = rawTargetPath.replace('file:/', '')
	
        //Below Gathers Information on SourceMap and SourceNode, then gathers a list of the names of open maps
	//Below is code to define source node map path and source node ID 
        //spaces are preserved as spaces rather than '%20', but other special characters might cause issues with the current code - although tested and Freeplane won't let you save maps with special characters in the filename
	def LinktoSourceNode = node.link.text.replace('%20', ' ')
	def nLnk = LinktoSourceNode.split('#')
	def SourceMapPath = nLnk[0].replace('freeplane:/%20/', '').replace('file:///','').replace('file:/','')

    //Below defines the Source Map's Name using its path
        //NB I think this should always work, but maybe there are situations where it won't?
    def mapPathLength = SourceMapPath.length()
    def mapNameStart = SourceMapPath.lastIndexOf("/")+1 //gets mindmap file name without path
    def SourceMapName = SourceMapPath.getAt(mapNameStart..(mapPathLength-4)) //removes '.mm'
    def SourceNodeID = nLnk[1]
    
    //Below searches TargetNode for any children with the same ID as SourcNode - i.e. whether the script has been run on this particular TargetNode before. If this is the case, the old/existing child is deleted because it will be updated with the newer one we are about to paste from SourceMap
    def childnode = node.find{ it.getId().contains(SourceNodeID) }
    if(childnode){
	    c.select(childnode)
        c.selected.delete()
    }
    
    //Below gathers a list of open maps, Set-Up for deciding behaviour after copying
    def OpenMaps = c.getOpenMindMaps() //Lists Open Mindmaps
    def OpenMapsNames = OpenMaps.collect{it.Name} //Gets names of Open Mindmaps
    def SourceMapIsOpen = OpenMapsNames.contains((SourceMapName)) //=true if SourceMap is already open
    
    //Opens SourceMap, Selects SourceNode, and copies it
	def SourceMap = c.mapLoader(SourceMapPath).withView().load()
	def SourceNode = SourceMap.node(SourceNodeID)
	c.select(SourceNode)
    menuUtils.executeMenuItems(['CopyAction'])
    
    if (SourceMapIsOpen == true){
        	//If SourceMap was already open, simply Switches back to TargetMap and Pastes SourceNode
	   def TargetMap = c.mapLoader(TargetMapPath).withView().load()
	   def TargetNode = TargetMap.node(TargetNodeID)
	   menuUtils.executeMenuItems(['PasteAction'])
    }
    else {
        //If SourceMap was not already open, then closes the instance of it we just opened in the background before pasting to TargetNode 
        //NB assumes that closing SourceMap automatically takes us to TargetMap - maybe there are cases where this is untrue?
        menuUtils.executeMenuItems(['CloseAction'])
        menuUtils.executeMenuItems(['PasteAction'])
    }
    }

Thanks so, so much to all of you for your contributions. Hopefully this will be useful to others! 

I'm assuming I can just change the ExecutionMode to all nodes so that I can sync every node in the map with a link - maybe this will be slow for large maps but hopefully not too bad.

**Reply by abc16361** [reply, marked as answer]
I just tested the method `appendBranch` and it works better, so I agree to use it. Copying does preserve node's id, but in the case that there are multiple target nodes linking to a same source node, that node id cannot be used to identify other node copies, so adding an attribute helps in this way.  Here I edit a little bit more and use headless loader, maybe it runs faster in a large map, and it works no matter whether the source map is open or not. In the new code, the whole map is scanned to update multiple nodes ( from the same source, NOT different source nodes ).

```
// @ExecutionModes({ON_SINGLE_NODE})


def nodeLink = node.link.text
def nLnk = nodeLink.split('#')
def mapPath = nLnk[0].replace('freeplane:/%20/', '').replace('file:///','').replace('file:/','')
def nid = nLnk[1]
def childnodes = c.find{ it.attributes.getNames().contains(nid) }
def childnode = node.find{ it.attributes.getNames().contains(nid) }
if(childnodes == []){
	def sourceMap = c.mapLoader(mapPath).load()
	def sourceNode = sourceMap.node(nid)
	sourceNode[nid] = ''
	c.selected.appendBranch(sourceNode)
	sourceNode.attributes.removeAll(nid)
	menuUtils.executeMenuItems(['SaveAll'])
	return
}
if(childnode == []){
	def sourceMap = c.mapLoader(mapPath).load()
	def sourceNode = sourceMap.node(nid)
	sourceNode[nid] = ''
	node.appendBranch(sourceNode)
	for( cd in childnodes){ 
		c.select(cd)
		cd.delete()
		c.selected.appendBranch(sourceNode)
	}
	sourceNode.attributes.removeAll(nid)
	menuUtils.executeMenuItems(['SaveAll'])
	return	
}
else{ 
	def sourceMap = c.mapLoader(mapPath).load()
	def sourceNode = sourceMap.node(nid)
	sourceNode[nid] = ''
	for( cd in childnodes){ 
		c.select(cd)
		cd.delete()
		c.selected.appendBranch(sourceNode)
	}
	sourceNode.attributes.removeAll(nid)
	menuUtils.executeMenuItems(['SaveAll'])
}
```

**Reply by CD80-86** [reply]
@abc16361 Just wanted to confirm that yes, copy and paste had a lot of issues when trying to apply the script on more than one node, in addition to what you are describing. For example, running the script I posted earlier recursively on several nodes would paste all the nodes onto one target rather than their respective correct targets.

I rewrote the script for my specific use case using appendBranch() and with a few if-else statements about exception cases (like nodes that don't have a link) I can now execute it recursively on all the nodes in my map, regardless of the Source they are linking to. Finished!

Your implementation is definitely neater and I will mark it as the Answer so that others can benefit best from it.

Thank you for all the help!

**Comment by EdoFro** [comment]
> > TL;DR is one big map would be too slow due to a large number of pasted images in the map.
> 
> Good to know that images hurt performance. My map doesn't have this amount of images, so I never noticed this problem.
> 
> >

There is an option in the Preferences dialog that says something like:

- “images cache” : Disk or Memory 

if you use the “Disk” option, the images are loaded to memory only when you unfold the branches that have the images and they are shown for the first time. 

This way you can manage big mindmaps with lot of images, just by folding the branches you are not longer using. 

I use this option since long time ago when RAM capacity was more of an issue, and I feel it is fast enough for not noticing that the images load only when needed. 

hope it helps
edo

**Reply by CD80-86** [reply]
I wasn't aware of this - thank you! I'll definitely keep this in mind if I decide to try switching to a single-map setup. I'm mainly reluctant to do so because it would involve a fair bit of copy-pasting across all my maps haha. But I will if I have to.
